File: api/shared/utils/schemaValidator.py
"""
Python Schema Validator Utility
Centralized validation using JSON schemas
Part of systematic resolution for validation centralization
"""

import json
import os
from typing import Dict, Any, List, Optional
from jsonschema import validate, ValidationError as JSONSchemaValidationError, Draft7Validator


class SchemaValidator:
    """Centralized schema validator for all entities"""
    
    def __init__(self):
        self.schemas = {}
        self.schema_dir = os.path.join(os.path.dirname(__file__), "..", "..", "..", "shared", "schemas")
        self._load_schemas()
    
    def _load_schemas(self):
        """Load all JSON schemas from the schemas directory"""
        try:
            # For now, use fallback schemas to avoid JSON reference resolution issues
            self._load_fallback_schemas()
            return
            
            schema_files = ['base.json', 'prompt.json', 'collection.json', 'playbook.json', 'user.json', 'cost.json']
            
            for schema_file in schema_files:
                schema_path = os.path.join(self.schema_dir, schema_file)
                if os.path.exists(schema_path):
                    with open(schema_path, 'r') as f:
                        schema_name = schema_file.replace('.json', '')
                        self.schemas[schema_name] = json.load(f)
                        
        except Exception as e:
            print(f"Warning: Could not load schemas: {e}")
            # Provide basic fallback schemas
            self._load_fallback_schemas()
    
    def _load_fallback_schemas(self):
        """Load basic fallback schemas if files are not found"""
        self.schemas = {
            'prompt': {
                'type': 'object',
                'required': ['name', 'content'],
                'properties': {
                    'name': {'type': 'string', 'minLength': 1},
                    'content': {'type': 'string', 'minLength': 1},
                    'description': {'type': 'string'},
                    'tags': {'type': 'array', 'items': {'type': 'string'}},
                    'variables': {'type': 'array'},
                    'id': {'type': 'string'},
                    'userId': {'type': 'string'},
                    'createdAt': {'type': 'string'},
                    'updatedAt': {'type': 'string'},
                    'isPublic': {'type': 'boolean'},
                    'category': {'type': 'string'},
                    'usageCount': {'type': 'integer', 'minimum': 0},
                    'version': {'type': 'string'}
                },
                'additionalProperties': True
            },
            'collection': {
                'type': 'object',
                'required': ['name'],
                'properties': {
                    'name': {'type': 'string', 'minLength': 1},
                    'description': {'type': 'string'},
                    'type': {'type': 'string', 'enum': ['private', 'shared_team', 'public_marketplace']},
                    'tags': {'type': 'array', 'items': {'type': 'string'}},
                    'id': {'type': 'string'},
                    'userId': {'type': 'string'},
                    'createdAt': {'type': 'string'},
                    'updatedAt': {'type': 'string'}
                },
                'additionalProperties': True
            },
            'playbook': {
                'type': 'object',
                'required': ['name', 'steps'],
                'properties': {
                    'name': {'type': 'string', 'minLength': 1},
                    'description': {'type': 'string'},
                    'steps': {'type': 'array', 'minItems': 1},
                    'variables': {'type': 'array'},
                    'id': {'type': 'string'},
                    'userId': {'type': 'string'},
                    'createdAt': {'type': 'string'},
                    'updatedAt': {'type': 'string'}
                },
                'additionalProperties': True
            },
            'user': {
                'type': 'object',
                'required': ['id', 'email'],
                'properties': {
                    'id': {'type': 'string'},
                    'email': {'type': 'string', 'format': 'email'},
                    'name': {'type': 'string'},
                    'role': {'type': 'string', 'enum': ['user', 'admin', 'guest']},
                    'createdAt': {'type': 'string'},
                    'updatedAt': {'type': 'string'}
                },
                'additionalProperties': True
            },
            'cost': {
                'type': 'object',
                'required': ['userId', 'amount'],
                'properties': {
                    'userId': {'type': 'string'},
                    'amount': {'type': 'number', 'minimum': 0},
                    'currency': {'type': 'string'},
                    'provider': {'type': 'string'},
                    'timestamp': {'type': 'string'},
                    'requestType': {'type': 'string'}
                },
                'additionalProperties': True
            }
        }
    
    def validate_entity(self, data: Dict[str, Any], entity_type: str) -> Dict[str, Any]:
        """
        Validate an entity against its schema
        
        Args:
            data: The data to validate
            entity_type: The type of entity (prompt, collection, playbook, user, etc.)
            
        Returns:
            Dictionary with 'valid' boolean and 'errors' list
        """
        if entity_type not in self.schemas:
            return {
                'valid': False,
                'errors': [f'Unknown entity type: {entity_type}']
            }
        
        schema = self.schemas[entity_type]
        
        try:
            validate(instance=data, schema=schema, cls=Draft7Validator)
            return {
                'valid': True,
                'errors': []
            }
        except JSONSchemaValidationError as e:
            return {
                'valid': False,
                'errors': [str(e)]
            }
        except Exception as e:
            return {
                'valid': False,
                'errors': [f'Validation error: {str(e)}']
            }
    
    def validate_batch(self, entities: List[Dict[str, Any]], entity_type: str) -> Dict[str, Any]:
        """
        Validate multiple entities of the same type
        
        Args:
            entities: List of entities to validate
            entity_type: The type of entity
            
        Returns:
            Dictionary with overall validation result and individual errors
        """
        results = []
        all_valid = True
        
        for i, entity in enumerate(entities):
            result = self.validate_entity(entity, entity_type)
            results.append({
                'index': i,
                'valid': result['valid'],
                'errors': result['errors']
            })
            if not result['valid']:
                all_valid = False
        
        return {
            'valid': all_valid,
            'results': results,
            'total_count': len(entities),
            'valid_count': sum(1 for r in results if r['valid'])
        }


# Global validator instance
_validator = None


def get_validator() -> SchemaValidator:
    """Get the global validator instance"""
    global _validator
    if _validator is None:
        _validator = SchemaValidator()
    return _validator


def validate_entity(data: Dict[str, Any], entity_type: str) -> Dict[str, Any]:
    """
    Convenience function to validate a single entity
    
    Args:
        data: The data to validate
        entity_type: The type of entity
        
    Returns:
        Dictionary with 'valid' boolean and 'errors' list
    """
    return get_validator().validate_entity(data, entity_type)


def validate_batch(entities: List[Dict[str, Any]], entity_type: str) -> Dict[str, Any]:
    """
    Convenience function to validate multiple entities
    
    Args:
        entities: List of entities to validate
        entity_type: The type of entity
        
    Returns:
        Dictionary with overall validation result
    """
    return get_validator().validate_batch(entities, entity_type)


# Export commonly used functions
__all__ = [
    'SchemaValidator',
    'get_validator',
    'validate_entity',
    'validate_batch'
]

File: shared/utils/schemaValidator.ts
/**
 * Schema Validator Utility - Simplified Version
 * Centralized validation for Sutra data models
 * Part of the systematic resolution for validation fragmentation
 */

export interface ValidationResult {
  isValid: boolean;
  errors: string[];
  data?: any;
}

export interface ValidationError {
  field: string;
  message: string;
  value?: any;
}

/**
 * Basic field validation functions
 */
export const validators = {
  required: (value: any, fieldName: string): string | null => {
    if (value === undefined || value === null || value === "") {
      return `${fieldName} is required`;
    }
    return null;
  },

  string: (
    value: any,
    fieldName: string,
    options: { minLength?: number; maxLength?: number } = {},
  ): string | null => {
    if (typeof value !== "string") {
      return `${fieldName} must be a string`;
    }
    if (options.minLength && value.length < options.minLength) {
      return `${fieldName} must be at least ${options.minLength} characters`;
    }
    if (options.maxLength && value.length > options.maxLength) {
      return `${fieldName} must be no more than ${options.maxLength} characters`;
    }
    return null;
  },

  email: (value: any, fieldName: string): string | null => {
    if (typeof value !== "string") {
      return `${fieldName} must be a string`;
    }
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    if (!emailRegex.test(value)) {
      return `${fieldName} must be a valid email address`;
    }
    return null;
  },

  enum: (
    value: any,
    fieldName: string,
    allowedValues: any[],
  ): string | null => {
    if (!allowedValues.includes(value)) {
      return `${fieldName} must be one of: ${allowedValues.join(", ")}`;
    }
    return null;
  },

  number: (
    value: any,
    fieldName: string,
    options: { min?: number; max?: number } = {},
  ): string | null => {
    if (typeof value !== "number" || isNaN(value)) {
      return `${fieldName} must be a number`;
    }
    if (options.min !== undefined && value < options.min) {
      return `${fieldName} must be at least ${options.min}`;
    }
    if (options.max !== undefined && value > options.max) {
      return `${fieldName} must be no more than ${options.max}`;
    }
    return null;
  },

  integer: (
    value: any,
    fieldName: string,
    options: { min?: number; max?: number } = {},
  ): string | null => {
    if (!Number.isInteger(value)) {
      return `${fieldName} must be an integer`;
    }
    return validators.number(value, fieldName, options);
  },

  boolean: (value: any, fieldName: string): string | null => {
    if (typeof value !== "boolean") {
      return `${fieldName} must be a boolean`;
    }
    return null;
  },

  array: (
    value: any,
    fieldName: string,
    options: { minItems?: number; maxItems?: number } = {},
  ): string | null => {
    if (!Array.isArray(value)) {
      return `${fieldName} must be an array`;
    }
    if (options.minItems && value.length < options.minItems) {
      return `${fieldName} must have at least ${options.minItems} items`;
    }
    if (options.maxItems && value.length > options.maxItems) {
      return `${fieldName} must have no more than ${options.maxItems} items`;
    }
    return null;
  },

  objectId: (value: any, fieldName: string): string | null => {
    if (typeof value !== "string" || value.length === 0) {
      return `${fieldName} must be a non-empty string`;
    }
    return null;
  },

  timestamp: (value: any, fieldName: string): string | null => {
    if (typeof value === "string") {
      const date = new Date(value);
      if (isNaN(date.getTime())) {
        return `${fieldName} must be a valid ISO timestamp`;
      }
    } else if (!(value instanceof Date)) {
      return `${fieldName} must be a Date object or ISO timestamp string`;
    }
    return null;
  },
};

/**
 * Schema definitions for each entity type
 */
export const schemas = {
  prompt: {
    required: ["id", "title", "description", "content", "userId", "createdAt"],
    fields: {
      id: { validator: "objectId" },
      title: { validator: "string", options: { minLength: 1, maxLength: 200 } },
      description: {
        validator: "string",
        options: { minLength: 1, maxLength: 1000 },
      },
      content: { validator: "string", options: { minLength: 1 } },
      userId: { validator: "objectId" },
      tags: { validator: "array", options: { maxItems: 10 } },
      isPublic: { validator: "boolean" },
      category: {
        validator: "enum",
        options: [
          "writing",
          "coding",
          "analysis",
          "creative",
          "business",
          "other",
        ],
      },
      createdAt: { validator: "timestamp" },
      updatedAt: { validator: "timestamp" },
      usageCount: { validator: "integer", options: { min: 0 } },
      version: { validator: "string" },
    },
  },

  collection: {
    required: ["id", "name", "description", "userId", "createdAt"],
    fields: {
      id: { validator: "objectId" },
      name: { validator: "string", options: { minLength: 1, maxLength: 100 } },
      description: {
        validator: "string",
        options: { minLength: 1, maxLength: 500 },
      },
      userId: { validator: "objectId" },
      prompts: { validator: "array" },
      isPublic: { validator: "boolean" },
      tags: { validator: "array", options: { maxItems: 10 } },
      category: {
        validator: "enum",
        options: ["personal", "team", "public", "template", "other"],
      },
      createdAt: { validator: "timestamp" },
      updatedAt: { validator: "timestamp" },
      promptCount: { validator: "integer", options: { min: 0 } },
    },
  },

  playbook: {
    required: ["id", "name", "description", "userId", "createdAt", "steps"],
    fields: {
      id: { validator: "objectId" },
      name: { validator: "string", options: { minLength: 1, maxLength: 100 } },
      description: {
        validator: "string",
        options: { minLength: 1, maxLength: 1000 },
      },
      userId: { validator: "objectId" },
      steps: { validator: "array", options: { minItems: 1 } },
      tags: { validator: "array", options: { maxItems: 10 } },
      category: {
        validator: "enum",
        options: [
          "automation",
          "analysis",
          "content",
          "customer_service",
          "development",
          "other",
        ],
      },
      isPublic: { validator: "boolean" },
      version: { validator: "string" },
      createdAt: { validator: "timestamp" },
      updatedAt: { validator: "timestamp" },
      executionCount: { validator: "integer", options: { min: 0 } },
      avgExecutionTime: { validator: "number", options: { min: 0 } },
    },
  },

  user: {
    required: ["id", "email", "role", "createdAt"],
    fields: {
      id: { validator: "objectId" },
      email: { validator: "email" },
      name: { validator: "string", options: { minLength: 1, maxLength: 100 } },
      role: { validator: "enum", options: ["admin", "user", "guest"] },
      permissions: { validator: "array" },
      createdAt: { validator: "timestamp" },
      updatedAt: { validator: "timestamp" },
      lastLoginAt: { validator: "timestamp" },
      isActive: { validator: "boolean" },
      emailVerified: { validator: "boolean" },
    },
  },

  cost: {
    required: [
      "id",
      "userId",
      "provider",
      "model",
      "requestId",
      "timestamp",
      "cost",
    ],
    fields: {
      id: { validator: "objectId" },
      userId: { validator: "objectId" },
      provider: {
        validator: "enum",
        options: ["openai", "anthropic", "google", "local"],
      },
      model: { validator: "string" },
      requestId: { validator: "objectId" },
      promptId: { validator: "objectId" },
      playbookId: { validator: "objectId" },
      timestamp: { validator: "timestamp" },
      cost: { validator: "number", options: { min: 0 } },
      inputTokens: { validator: "integer", options: { min: 0 } },
      outputTokens: { validator: "integer", options: { min: 0 } },
      totalTokens: { validator: "integer", options: { min: 0 } },
      requestDuration: { validator: "number", options: { min: 0 } },
      requestType: {
        validator: "enum",
        options: [
          "prompt_execution",
          "playbook_step",
          "test_execution",
          "api_call",
        ],
      },
      status: {
        validator: "enum",
        options: ["success", "error", "timeout", "cancelled"],
      },
      errorMessage: { validator: "string" },
    },
  },
};
/**
 * Validate data against a schema
 */
export function validateData(
  schemaName: keyof typeof schemas,
  data: any,
  partial = false,
): ValidationResult {
  const schema = schemas[schemaName];
  if (!schema) {
    return {
      isValid: false,
      errors: [`Unknown schema: ${schemaName}`],
    };
  }

  const errors: string[] = [];
  const validatedData: any = {};

  // Check required fields (only for non-partial validation)
  if (!partial) {
    for (const requiredField of schema.required) {
      if (!(requiredField in data)) {
        errors.push(`Missing required field: ${requiredField}`);
        continue;
      }
    }
  }

  // Validate each field that exists in the data
  for (const [fieldName, fieldValue] of Object.entries(data)) {
    const fieldSchema = schema.fields[fieldName];

    if (!fieldSchema) {
      // Field not in schema - skip or warn?
      continue;
    }

    const validatorName = fieldSchema.validator;
    const validatorOptions = fieldSchema.options;

    if (validatorName in validators) {
      const validator = validators[validatorName as keyof typeof validators];
      let error: string | null = null;

      // Handle different validator signatures
      if (validatorName === "enum" && validatorOptions) {
        error = (validator as any)(fieldValue, fieldName, validatorOptions);
      } else if (validatorOptions) {
        error = (validator as any)(fieldValue, fieldName, validatorOptions);
      } else {
        error = (validator as any)(fieldValue, fieldName);
      }

      if (error) {
        errors.push(error);
      } else {
        validatedData[fieldName] = fieldValue;
      }
    }
  }

  return {
    isValid: errors.length === 0,
    errors,
    data: errors.length === 0 ? validatedData : undefined,
  };
}

/**
 * Convenience validation functions
 */
export const validatePrompt = (data: any, partial = false) =>
  validateData("prompt", data, partial);
export const validateCollection = (data: any, partial = false) =>
  validateData("collection", data, partial);
export const validatePlaybook = (data: any, partial = false) =>
  validateData("playbook", data, partial);
export const validateUser = (data: any, partial = false) =>
  validateData("user", data, partial);
export const validateCost = (data: any, partial = false) =>
  validateData("cost", data, partial);

/**
 * Create validation middleware for Express.js
 */
export function createValidationMiddleware(
  schemaName: keyof typeof schemas,
  options: {
    validateBody?: boolean;
    validateQuery?: boolean;
    partial?: boolean;
  } = {},
) {
  return (req: any, res: any, next: any) => {
    const errors: string[] = [];

    if (options.validateBody && req.body) {
      const result = validateData(schemaName, req.body, options.partial);
      if (!result.isValid) {
        errors.push(`Body validation: ${result.errors.join(", ")}`);
      } else if (result.data) {
        req.body = result.data; // Use validated data
      }
    }

    if (options.validateQuery && req.query) {
      const result = validateData(schemaName, req.query, true); // Query params are always partial
      if (!result.isValid) {
        errors.push(`Query validation: ${result.errors.join(", ")}`);
      }
    }

    if (errors.length > 0) {
      return res.status(400).json({
        error: "Validation failed",
        details: errors,
      });
    }

    next();
  };
}

/**
 * Validation middleware creators
 */
export const createPromptValidation = (options = {}) =>
  createValidationMiddleware("prompt", options);
export const createCollectionValidation = (options = {}) =>
  createValidationMiddleware("collection", options);
export const createPlaybookValidation = (options = {}) =>
  createValidationMiddleware("playbook", options);
export const createUserValidation = (options = {}) =>
  createValidationMiddleware("user", options);
export const createCostValidation = (options = {}) =>
  createValidationMiddleware("cost", options);
